### ISSUE TITLE
Creation of file with flag="a" and driver="mpio" not working

### ISSUE THREAD

**Issue by sbacchio** [h5py/h5py#1920]
Hello,

I'm running in a issue introduced by PR #1902 (https://github.com/h5py/h5py/pull/1902/files#diff-6515aaa459980029aa5ae782298c82cacb68f8f6f3218a9f0419f7148bd80f30L208).
When opening a non existing file with `File(filename, "a", driver="mpio", comm=comm)` `OSError` gets raised.
This is because `mpi4py` manages the opening and it raises `OSError` instead of `FileNotFoundError`.

Here an error log
```
  /opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/h5py/_hl/files.py:211: in make_fid
      with File(filename, "a", driver="mpio", comm=comm) as h5f:
  /opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/h5py/_hl/files.py:444: in __init__
      fid = make_fid(name, mode, userblock_size,
  /opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/h5py/_hl/files.py:211: in make_fid
      fid = h5f.open(name, h5f.ACC_RDWR, fapl=fapl)
  h5py/_objects.pyx:54: in h5py._objects.with_phil.wrapper
      fid = h5f.open(name, h5f.ACC_RDWR, fapl=fapl)
  h5py/_objects.pyx:54: in h5py._objects.with_phil.wrapper
      ???
      ???
  h5py/_objects.pyx:55: in h5py._objects.with_phil.wrapper
  h5py/_objects.pyx:55: in h5py._objects.with_phil.wrapper
      ???
  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
  
      ???
  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
  
  >   ???
  E   OSError: Unable to open file (MPI_ERR_NO_SUCH_FILE: no such file or directory)
  
  h5py/h5f.pyx:100: OSError
  >   ???
  E   OSError: Unable to open file (MPI_ERR_NO_SUCH_FILE: no such file or directory)
  
  h5py/h5f.pyx:100: OSError
```

**Comment by sbacchio** [comment]
I see it's connected to issue #1916 and the solution proposed by @takluyver (https://github.com/h5py/h5py/issues/1916#issuecomment-872216056).

But why not restoring `OSError`?

**Comment by takluyver** [comment]
*Sigh* Maybe we need to expect the more specific FileNotFoundError only when we're using the default file driver, and go back to catching the generic OSError on other drivers. :disappointed:

**Comment by sbacchio** [comment]
Yes, I would be in favor of that solution. The flag "a" is very useful and critical for flexibility.

**Comment by takluyver** [comment]
It's not critical for anything - the code using h5py could easily achieve the same thing by trying with `r+` mode first, and if that fails switching to `x` mode. However, as it's there, it should work.

Do you want to have a go at making a pull request for that?

**Comment by sbacchio** [comment]
Sure I can do it

**Comment by sbacchio** [comment]
do you have tests for the drivers in order to make sure that it will not break again?

**Comment by takluyver** [comment]
I think append mode is only tested with the default driver (or we'd have caught this already), but here are some relevant tests you can start with:

https://github.com/h5py/h5py/blob/cc34c585ab29e859abda1d99830e7b9868197adb/h5py/tests/test_file.py#L83-L99

https://github.com/h5py/h5py/blob/cc34c585ab29e859abda1d99830e7b9868197adb/h5py/tests/test_file.py#L228-L316

**Comment by takluyver** [comment]
And the MPI tests:

https://github.com/h5py/h5py/blob/cc34c585ab29e859abda1d99830e7b9868197adb/h5py/tests/test_file.py#L755-L783

**Comment by sbacchio** [comment]
@takluyver created PR #1922

**Comment by takluyver** [comment]
Closed by #1922
